src/toil/batchSystems/awsBatch.py

```diff
@@ -0,0 +1,519 @@
+                job['userScript'] = self.user_script
+            # Encode it in a form we can send in a command-line argument. Pickle in
+            # the highest protocol to prevent mixed-Python-version workflows from
+            # trying to work. Make sure it is text so we can ship it to Kubernetes
+            # via JSON.
+            encoded_job = base64.b64encode(pickle.dumps(job, pickle.HIGHEST_PROTOCOL)).decode('utf-8')
+            # Make a command to run it in the exacutor
+            command_list = ['_toil_contained_executor', encoded_job]
+
+            # Compose a job spec to submit
+            job_spec = {
+                'jobName': job_name,
+                'jobQueue': self.queue,
+                'jobDefinition': self.__get_or_create_job_definition(),
+                'containerOverrides': {
+                    'command': command_list,
+                    'environment': [{'name': k, 'value': v} for k, v in environment.items()],
+                    'resourceRequirements': [
+                        {'type': 'MEMORY', 'value': str(math.ceil(max(4, to_mib(job_desc.memory))))},  # A min of 4 is demanded by the API
+                        {'type': 'VCPU', 'value': str(math.ceil(job_desc.cores))}  # Must be at least 1 on EC2, probably can't be 1.5
+                    ]
+                }
+            }
+            if self.owner_tag:
+                # We are meant to tag everything with an owner
+                job_spec['tags'] = {'Owner': self.owner_tag}
+
+            # Launch it and get back the AWS ID that we can use to poll the task.
+            # TODO: retry!
+            response = self.client.submit_job(**job_spec)
+            aws_id = response['jobId']
+
+            # Tie it to the numeric ID
+            self.bs_id_to_aws_id[bs_id] = aws_id
+            self.aws_id_to_bs_id[aws_id] = bs_id
+
+            logger.debug('Launched job: %s', job_name)
+
+            return bs_id
+
+    def __ensafen_name(self, input_name: str) -> str:
+        """
+        Make a job name safe for Amazon Batch.
+        From the API docs:
+
+            It can be up to 128 letters long. The first character must be
+            alphanumeric, can contain uppercase and lowercase letters, numbers,
+            hyphens (-), and underscores (_).
+        """
+        # Do replacements to enhance readability
+        input_name = input_name.replace(" ", "-")
+        # Keep only acceptable characters
+        kept_chars = [c for c in input_name if c.isalnum() or c == '-' or c == '_']
+        if len(kept_chars) == 0 or not kept_chars[0].isalnum():
+            # Make sure we start with something alphanumeric
+            kept_chars = ['j'] + kept_chars
+        # Keep no more than the limit of them
+        kept_chars = kept_chars[:128]
+        # And re-compose them into a string
+        return ''.join(kept_chars)
+
+    def __get_runtime(self, job_detail: Dict[str, Any]) -> Optional[float]:
```

Comment:
I don't think we're going to be extending this class, so name mangling as a tool to avoid collisions doesn't really help us here.

But I do mean this method to be at class scope. I don't *want* this method to be re-usable elsewhere unless it is officially promoted to part of the API (and given a test, etc.). At that point we can rename it with find/replace because... it won't be mentioned anywhere else in the codebase. I could use some kind of custom decorator that audits the call stack to make sure nobody is calling it who shouldn't be, but the two underscores is standard, easier, and less obnoxious if someone *does* have to hack around it for a good reason.

I think when I was refactoring the leader I got mad about how the particular behavior of single-underscore-prefixed stuff had grown to matter in other parts of the codebase. I wanted to be able to change the semantics of or drop a thing that started with an underscore, but it turned out it was being used elsewhere or under test in the unit tests. The `AbstractJobStore` even has e.g. `_read_from_url()` as part of its defined API, which concrete job store classes are required to implement.

Maybe if I was using a proper IDE that knew about all the project files at once, I wouldn't trip over these usages because the IDE would be surfacing them for me?

Anyway, now when I add something that I don't want to grow secret connections to other parts of the codebase over time, I put two underscores on it.

Reply:
PEP-8 recommends using __ only in base classes, to keep subclasses out of your hair, and that _ ought to be enough to discourage people in other parts of your codebase.

Maybe I will change to _ and put in the docstrings that nobody should call these? And maybe we need to make it a review check that nobody is calling _-prefixed things they shouldn't be?